[DataLoader] Implementing communication processes for Map-style DataPipes

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68549



cc @SsnL @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32922676](https://our.internmc.facebook.com/intern/diff/D32922676)

NivekT committed:
[DataLoader] Implementing communication processes for Map-style DataPipes

[ghstack-poisoned]

NivekT commented on test/test_dataloader.py:
```diff
@@ -2033,6 +2034,24 @@ def clean_me(process, req_queue, res_queue):
+        actual = list(local_datapipe)
+        clean_me(process, req_queue, res_queue)
+
+        self.assertEqual([(x, x) for x in range(100)], actual)
```
Worth noting that currently `getitem` returns both `(key, value)`. Should we just return `value`?

NivekT commented on torch/utils/data/communication/map.py:
```diff
@@ -0,0 +1,143 @@
+                except NotAvailable:
+                    yield True
+                    continue
+                except IndexError as e:
```
I choose to trigger `StopIteration` when an invalid index is provided. This works well when the loop incrementally passes in indices and eventually goes out of bound. 

Do you think an alternative of raising IndexError and handling that is better?

NivekT commented on torch/utils/data/communication/protocol.py:
```diff
@@ -62,80 +62,126 @@ def get_new_request(self, block=False):
         self._req_received = None
 
-    def response_terminate(self):
+    def response_stop_iteration(self):
```
This handles the case when index goes out of bound.

VitalyFedyunin replied:
Please name it differently

NivekT commented on torch/utils/data/dataloader_experimental.py:
```diff
@@ -55,6 +55,13 @@ def clean_me(thread, req_queue, res_queue):
             clean_me(thread, req_queue, res_queue)
 
 
+def sharding_worker_init_fn(worker_init_fn, worker_id):
```
I moved this inner function out to allow pickling.

NivekT committed:
Update on "[DataLoader] Implementing communication processes for Map-style DataPipes"




cc @SsnL @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/dataloader_experimental.py:
```diff
@@ -33,7 +33,7 @@ def __iter__(self):
                 if dp not in exclude_datapipes:
                     try:
-                        value = dp.nonblocking_next()
+                        value = dp.nonblocking_next()  # TODO: call nonblocking_getitem(index) for MapDataPipe
```
Drop this comment

VitalyFedyunin reviewed (approved):
Overall looks good

NivekT committed:
Update on "[DataLoader] Implementing communication processes for Map-style DataPipes"




cc @SsnL @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).

NivekT committed:
Update on "[DataLoader] Implementing communication processes for Map-style DataPipes"




cc @SsnL @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32922676](https://our.internmc.facebook.com/intern/diff/D32922676)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).

NivekT committed:
Update on "[DataLoader] Implementing communication processes for Map-style DataPipes"




cc @SsnL @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32922676](https://our.internmc.facebook.com/intern/diff/D32922676)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32922676).
